packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx

```diff
@@ -0,0 +1,217 @@
+'use client';
+
+import * as React from 'react';
+import { useAnimationFrame, useEventCallback } from '@fluentui/react-utilities';
+import type { StaggerProps } from './stagger-types';
+import {
+  staggerItemsVisibilityAtTime,
+  type StaggerItemsVisibilityAtTimeParams,
+  DEFAULT_ITEM_DURATION,
+  type StaggerChildMapping,
+} from './utils';
+
+export interface UseStaggerItemsVisibilityParams
+  extends Pick<StaggerProps, 'onMotionFinish'>,
+    Omit<StaggerItemsVisibilityAtTimeParams, 'elapsed' | 'itemCount'> {
+  hideMode: StaggerProps['hideMode'];
+  childMapping: StaggerChildMapping;
+}
+
+/**
+ * Hook that tracks the visibility of a staggered sequence of items as time progresses.
+ *
+ * Behavior summary for all hide modes:
+ * - On the first render, items are placed in their final state (enter => visible, exit => hidden)
+ *   and no animation runs.
+ * - On subsequent renders when direction changes, items animate from the opposite state
+ *   to the final state over the stagger timeline.
+ * - Changes to the `reversed` prop do not trigger re-animation; they only affect the order
+ *   during the next direction change animation.
+ *
+ * This hook uses child key mapping instead of item count to track individual items.
+ * This allows it to correctly handle:
+ * - Items being added and removed simultaneously (when count stays the same)
+ * - Items being reordered
+ * - Individual item identity across renders
+ *
+ * @param childMapping - Mapping of child keys to elements and indices
+ * @param itemDelay - Milliseconds between the start of each item's animation
+ * @param itemDuration - Milliseconds each item's animation lasts
+ * @param direction - 'enter' (show items) or 'exit' (hide items)
+ * @param reversed - Whether to reverse the stagger order (last item first)
+ * @param onMotionFinish - Callback fired when the full stagger sequence completes
+ * @param hideMode - How children's visibility is managed: 'visibleProp', 'visibilityStyle', or 'unmount'
+ *
+ * @returns An object with `itemsVisibility: Record<string, boolean>` indicating which items are currently visible by key
+ */
+export function useStaggerItemsVisibility({
+  childMapping,
+  itemDelay,
+  itemDuration = DEFAULT_ITEM_DURATION,
+  direction,
+  reversed = false,
+  onMotionFinish,
+  hideMode = 'visibleProp',
+}: UseStaggerItemsVisibilityParams): { itemsVisibility: Record<string, boolean> } {
+  const [requestAnimationFrame, cancelAnimationFrame] = useAnimationFrame();
+
+  // Stabilize the callback reference to avoid re-triggering effects on every render
+  const stableOnMotionFinish = useEventCallback(
```

Comment:
```suggestion
  const handleOnMotionFinish = useEventCallback(
```

nit to be consistent with naming in other files